pkg/registry/core/service/storage/alloc.go

```diff
@@ -0,0 +1,1101 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/klog/v2"
+	apiservice "k8s.io/kubernetes/pkg/api/service"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/core/validation"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
+	"k8s.io/kubernetes/pkg/registry/core/service/portallocator"
+	netutils "k8s.io/utils/net"
+)
+
+// Allocators encapsulates the various allocators (IPs, ports) used in
+// Services.
+type Allocators struct {
+	serviceIPAllocatorsByFamily map[api.IPFamily]ipallocator.Interface
+	defaultServiceIPFamily      api.IPFamily // --service-cluster-ip-range[0]
+	serviceNodePorts            portallocator.Interface
+}
+
+// ServiceNodePort includes protocol and port number of a service NodePort.
+type ServiceNodePort struct {
+	// The IP protocol for this port. Supports "TCP" and "UDP".
+	Protocol api.Protocol
+
+	// The port on each node on which this service is exposed.
+	// Default is to auto-allocate a port if the ServiceType of this Service requires one.
+	NodePort int32
+}
+
+// This is a trasitionary function to facilitate service REST flattening.
+func makeAlloc(defaultFamily api.IPFamily, ipAllocs map[api.IPFamily]ipallocator.Interface, portAlloc portallocator.Interface) Allocators {
+	return Allocators{
+		defaultServiceIPFamily:      defaultFamily,
+		serviceIPAllocatorsByFamily: ipAllocs,
+		serviceNodePorts:            portAlloc,
+	}
+}
+
+func (al *Allocators) allocateCreate(service *api.Service, dryRun bool) (transaction, error) {
+	result := metaTransaction{}
+
+	// Ensure IP family fields are correctly initialized.  We do it here, since
+	// we want this to be visible even when dryRun == true.
+	if err := al.initIPFamilyFields(After{service}, Before{nil}); err != nil {
+		return nil, err
+	}
+
+	// Allocate ClusterIPs
+	//TODO(thockin): validation should not pass with empty clusterIP, but it
+	//does (and is tested!).  Fixing that all is a big PR and will have to
+	//happen later.
+	if txn, err := al.txnAllocClusterIPs(service, dryRun); err != nil {
+		result.Revert()
```

Comment:
I put it there for congruence with the second part. If, for some reason, it got rearranged, it would still be correct.  Happy to remove if it offends?  But that was my reasoning

Reply:
irrelevant now that the defer() is in.